integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/WindowFirstLast.java

```diff
@@ -0,0 +1,146 @@
+import com.querydsl.core.types.dsl.Expressions;
+import com.querydsl.core.types.dsl.SimpleExpression;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * {@code WindowFirstLast} is a builder for window function expressions.
+ * Analog to {@link com.querydsl.sql.WindowFirstLast}.
+ *
+ * @param <T> Result type of expression
+ * @author Jan-Willem Gmelig Meyling
+ * @since 1.5.0
+ */
+public class WindowFirstLast<T> extends MutableExpressionBase<T> {
+
+    private static final long serialVersionUID = 4107262569593794721L;
+
+    private static final String ORDER_BY = "order by ";
+
+    private final List<OrderSpecifier<?>> orderBy = new ArrayList<OrderSpecifier<?>>();
+
+    @Nullable
+    private transient volatile SimpleExpression<T> value;
+
+    private final Expression<T> target;
+
+    private final boolean first;
+
+    public WindowFirstLast(WindowOver<T> target, boolean first) {
+        super(target.getType());
+        this.target = target;
+        this.first = first;
+    }
+
+    @Override
+    public <R,C> R accept(Visitor<R,C> v, C context) {
+        return getValue().accept(v, context);
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expression
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(ComparableExpressionBase<?> orderBy) {
+        value = null;
+        this.orderBy.add(orderBy.asc());
+        return this;
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expressions
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(ComparableExpressionBase<?>... orderBy) {
+        value = null;
+        for (ComparableExpressionBase<?> e : orderBy) {
+            this.orderBy.add(e.asc());
+        }
+        return this;
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expression
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(OrderSpecifier<?> orderBy) {
+        value = null;
+        this.orderBy.add(orderBy);
+        return this;
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expressions
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(OrderSpecifier<?>... orderBy) {
+        value = null;
+        Collections.addAll(this.orderBy, orderBy);
+        return this;
+    }
+
+    SimpleExpression<T> getValue() {
+        if (value == null) {
+            if (orderBy.isEmpty()) {
+                // TODO this check should be static
+                throw new IllegalStateException("No order by arguments given");
```

Comment:
Is that really necessary? I think in SQL if you use an empty over clause like in `first_value(x) over()` it will use the ordering of the query.

Reply:
Window First/Last is removed. It was never supported.